README.md
# YouTube Timestamps

YouTube Timestamps is a plugin to track timestamps on any video which contains them.

On YouTube, a content creator can add timestamps to either the description, or within the comments - which any user is free and able to do if the comment section permits. This plugin will grab those timestamps and display them in an easy to read format. The current timestamp selected will be highlighted, and marked as a heading under the title of the video.

## Installing

Install via the Chrome Web Store
[YouTube Timestamps](https://chrome.google.com/webstore/detail/youtube-timestamps/mopckogdjleiljpnmmafmlieekoenhbe)

You can install this app by downloading/cloning this repository,
Unzip the folder, and load unpacked on Chrome.

## Running the tests

`npm test`

## Contributing

Most of the issues are quite simple, mainly to do with parsing or ensuring a timestamp is properly caught.
If you want to contribute, check out the [Issues](https://github.com/TylerJDev/YouTube_Timestamps/issues) tab, or head into the [Projects](https://github.com/TylerJDev/YouTube_Timestamps/projects) tab to see what's currently on the menu.

### Writing issues

Please stick with this format when writing an issue

**Occurrences**
- Where this issue happens; provide links or details on how to get to the issue causing page.

**Steps to reproduce**
- How to produce this issue; leave steps on how to produce this issue as mentioned.

**Expected result**
- What should be expected; What is the expected result.

**Actual Result**
- What is currently happening; List the end result of the issue here.

**Comments**
- Any comments or more details that wasn't mentioned above.


## Versioning
For future versioning updates, check the changelog.

1.0.4 - Added initial setting "color" to handleYoutube.js

1.0.3 - Removed console log(s), fixed issue where timestamp might be undefined

1.0.2 - Added another fix regarding plugin setting, changed string value to boolean.

1.0.1 - Fixed issue with plugin setting, added new icon

1.0.0 - Initial Release

0.3.0 - Added proper A11Y, improved color input with combobox functionality 

0.2.9 - Fixed issue #5 & improved parsing

0.2.8 - Fixed issue #13

0.2.8 - Fixed issue #11

0.2.7 - Fixed issue #7

0.2.6 - Fixed issue #4

0.2.5 - Added interface

0.1.5 - More cleaning

0.1.4 - Multiple bug fixes, cleaned up code, added more a11y

0.1.3 - Fixed issue #6, improved file organization

0.1.2 - Fixed issue #10 as well as an issue that caused timestamps grabbed from comments not to update

0.1.1 - Fixed issue #7 regarding timestamp HTML being displayed on videos without timestamps

0.1.0 - Added Next/Previous track functionality

## Future additions

* Ability to "Favorite" timestamp
* Option to show current timestamp in title
* Force load of timestamps

## Authors

Plugin was made by Tyler Jones, any contributors will be listed here.

## Acknowledgments

I'd like to thank my friend [Aaron](https://github.com/AaronChapman) for helping me stay committed to this plugin.


CHANGELOG.md
## Change Log
Any changes made to this plugin should be noted in this file.

## 1.06 (4/28/2020)

## Changed
* Fixed issue with "Enable background" not working properly
* Added styling fix for "Above Title" setting

## 1.05 - (4/27/2020)

### Changed
* Adjusted UI style for timestamp indicator
* Added way to toggle old UI style
* Added a way to check if a video contains the new chapters feature, #19


js/tests/findTimestamp.test.js
const f_timestamps = require('../findTimestamp.js');

describe('Tesitng findTimestamp module', () => {
  beforeEach(() => {
    // Create "mock" element
    document.body.innerHTML =
    '<div id="description">' +
      '<div class="content">' +
        '<a href="/watch?v=123456=0s">00:00</a>' +
        '<a href="/watch?v=123456=0s">00:01</a>' +
        '<a href="/watch?v=123456=3s">00:03</a>' +
        '<a href="/watch?v=123456=3s">Nani!?</a>' +
        '<video class="video-stream"></video>' +
      '</div>' +
    '</div>';
  });

  test('Check currentTimestamp Times', () => {
    expect(f_timestamps.findCurrentTimestamps()['links_tracks'].length).toEqual(3);
  });
});

describe('Testing grabbing timestamp text', () => {
  beforeEach(() => {
    // Create "mock" element
    document.body.innerHTML =
    '<div id="description">' +
      '<div class="content">Sounds by Yoko Shimomura ^-^' +
        'Not all songs are in this video, but I hope you enjoy :3\n' +
        '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=0s">00:00</a> Otar Botkoveli Debate\n' +
        '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=246s">04:06</a> Prague - First visit ambience\n' +
        '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=453s">07:33</a> Prague - General ambience\n' +
        '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=453s">09:33</a> Prague - General ambience Part 2\n' +
        'Detroit City ambience<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=1833s">30:33</a>\n' +
        '<video class="video-stream"></video>' +
      '</div>' +
    '</div>'
  });

  test('Grab correct timestamp text', () => {
    expect(f_timestamps.findCurrentTimestamps()['links_tracks'][0][1]).toBe('Otar Botkoveli Debate');
    expect(f_timestamps.findCurrentTimestamps()['links_tracks'][4][1]).toBe('Detroit City ambience');
  });

  test('Grab current track', () => {
    expect(f_timestamps.findCurrentTimestamps(430)['current_track']).toBe('Prague - First visit ambience');
    expect(f_timestamps.findCurrentTimestamps(4300)['current_track']).toBe('Detroit City ambience');
    expect(f_timestamps.findCurrentTimestamps(0)['current_track']).toBe('Otar Botkoveli Debate');
  });
});

describe('Testing correct attributes on links', () => {
  beforeEach(() => {
    // Create "mock" element
    document.body.innerHTML =
    '<div id="description">' +
      '<div class="content">Sounds by Yoko Shimomura ^-^' +
        'Not all songs are in this video, but I hope you enjoy :3\n' +
        '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=0s">00:00</a> Otar Botkoveli Debate\n' +
        '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=246s">04:06</a> Prague - First visit ambience\n' +
        '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=453s">07:33</a> Prague - General ambience\n' +
        '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=453s">09:33</a> Prague - General ambience Part 2\n' +
        'Detroit City ambience<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=1833s">30:33</a>\n' +
        '<video class="video-stream"></video>' +
      '</div>' +
    '</div>';
  });

  test('Testing current class added to current timestamp link', () => {
    f_timestamps.findCurrentTimestamps(0);
    expect(document.querySelectorAll('a.selected_yt_timestamp_link').length).toBeTruthy();
    expect(document.querySelector('a.selected_yt_timestamp_link').textContent).toBe('00:00');

    f_timestamps.findCurrentTimestamps(430);
    expect(document.querySelectorAll('a.selected_yt_timestamp_link').length).toBe(1);
    expect(document.querySelector('a.selected_yt_timestamp_link').textContent).toBe('04:06');
  });
});


js/tests/handleYoutube.test.js
const mock = require('./mocks/MutationObserver.mock'); // @https://jestjs.io/docs/en/manual-mocks
const c_timestamps = require('../handleYoutube.js');

describe('Tesitng currentTimestamp module', () => {
  test('Check currentTimestamp Times', () => {
    expect(c_timestamps.timestampToSeconds('1:00')).toBe(60);
    expect(c_timestamps.timestampToSeconds('10:00')).toBe(600);
    expect(c_timestamps.timestampToSeconds('59:00')).toBe(3540);
    expect(c_timestamps.timestampToSeconds('2:00')).toBe(120);
    expect(c_timestamps.timestampToSeconds('01:00:00')).toBe(3600);
    expect(c_timestamps.timestampToSeconds('10:00:00')).toBe(36000);
    expect(c_timestamps.timestampToSeconds('00:00:00')).toBe(0);

    // Should be false, due to being more than 2 colons
    expect(c_timestamps.timestampToSeconds('00:00:00:00')).toBeFalsy();
    expect(c_timestamps.timestampToSeconds('Hello, World!')).toBeFalsy();
    expect(c_timestamps.timestampToSeconds('0000')).toBeFalsy();
    expect(c_timestamps.timestampToSeconds('1234')).toBeFalsy();
  });

  test('Get current timestamp time', () => {
    expect(c_timestamps.determineTimeSlot(3, ['0:00', '2:33', '3:56', '6:40', '11:23'])).toBe('0:00');
    expect(c_timestamps.determineTimeSlot(690, ['0:00', '2:33', '3:56', '6:40', '11:23'])).toBe('11:23');
    expect(c_timestamps.determineTimeSlot(20, ['0:05', '0:10', '0:19'])).toBe('0:19');
  })
});

describe('Tesitng findTimestamp module', () => {
  beforeEach(() => {
    // Create "mock" element
    document.body.innerHTML =
    `<div id="container">` +
      `<h1 class="title"></h1>` +
      '<div id="description">' +
        '<div class="content">' +
          '<a href="/watch?v=123456=0s">00:00</a>' +
          '<a href="/watch?v=123456=0s">00:01</a>' +
          '<a href="/watch?v=123456=3s">00:03</a>' +
          '<a href="/watch?v=123456=3s">Nani!?</a>' +
          '<video class="video-stream"></video>' +
        '</div>' +
      '</div>' +
    '</div>';
  });

  test('Check currentTimestamp Times', () => {
    expect(c_timestamps.findCurrentTimestamps()['links_tracks'].length).toEqual(3);
  });
});

describe('Testing grabbing timestamp text', () => {
  beforeEach(() => {
    // Create "mock" element
    document.body.innerHTML =
    `<div id="container">` +
      `<h1 class="title"></h1>` +
      '<div id="description">' +
        '<div class="content">Sounds by Yoko Shimomura ^-^' +
          'Not all songs are in this video, but I hope you enjoy :3\n' +
          '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=0s">00:00</a> Otar Botkoveli Debate\n' +
          '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=246s">04:06</a> Prague - First visit ambience\n' +
          '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=453s">07:33</a> Prague - General ambience\n' +
          '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=453s">09:33</a> Prague - General ambience Part 2\n' +
          'Detroit City ambience<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=1833s">30:33</a>\n' +
          '<video class="video-stream"></video>' +
        '</div>' +
      '</div>' +
    '</div>';
  });

  test('Grab correct timestamp text', () => {
    expect(c_timestamps.findCurrentTimestamps()['links_tracks'][0][1]).toBe('Otar Botkoveli Debate');
    expect(c_timestamps.findCurrentTimestamps()['links_tracks'][4][1]).toBe('Detroit City ambience');
  });

  test('Grab current track', () => {
    expect(c_timestamps.findCurrentTimestamps(430)['current_track']).toBe('Prague First visit ambience');
    expect(c_timestamps.findCurrentTimestamps(4300)['current_track']).toBe('Detroit City ambience');
    expect(c_timestamps.findCurrentTimestamps(0)['current_track']).toBe('Otar Botkoveli Debate');
  });
});

describe('Testing correct attributes on links', () => {
  beforeEach(() => {
    // Create "mock" element
    document.body.innerHTML =
    `<div id="container">` +
      `<h1 class="title"></h1>` +
      '<div id="description">' +
        '<div class="content">Sounds by Yoko Shimomura ^-^' +
          'Not all songs are in this video, but I hope you enjoy :3\n' +
          '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=0s">00:00</a> Otar Botkoveli Debate\n' +
          '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=246s">04:06</a> Prague - First visit ambience\n' +
          '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=453s">07:33</a> Prague - General ambience\n' +
          '<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=453s">09:33</a> Prague - General ambience Part 2\n' +
          'Detroit City ambience<a class="yt-simple-endpoint style-scope yt-formatted-string yt_timestamp_link" spellcheck="false" href="/watch?v=7lERcfsqJSk&amp;t=1833s">30:33</a>\n' +
          '<video class="video-stream"></video>' +
        '</div>' +
      '</div>' +
    '</div>';
  });

  test('Testing current class added to current timestamp link', () => {
    c_timestamps.findCurrentTimestamps(0);
    expect(document.querySelectorAll('a.selected_yt_timestamp_link').length).toBeTruthy();
    expect(document.querySelector('a.selected_yt_timestamp_link').textContent).toBe('00:00');

    c_timestamps.findCurrentTimestamps(430);
    expect(document.querySelectorAll('a.selected_yt_timestamp_link').length).toBe(1);
    expect(document.querySelector('a.selected_yt_timestamp_link').textContent).toBe('04:06');
  });
});


js/handleYoutube.js
/* eslint-disable no-nested-ternary */
/* eslint-disable max-len */
/* eslint-disable no-use-before-define */
const targetEle = document.querySelector('head > title');
const allowDebug = false;
const timestampHistory = { history: {} };

let timeStampObj = { links_tracks: [] };
let activeTimeUpdate = false;
let currentDescr;
let mutationSet = false;
let cContent = false;
let withinComments = '';
let settings = {
  toggle_heading: true, disable_background: true, below_title: true, above_title: false, color: '#575757',
};
let ytChapterContainer = [];

function consoleLogger(logLevel, ...args) {
  if (allowDebug) console[logLevel](...args); // eslint-disable-line
}

const targetObserve = new MutationObserver(() => {
  consoleLogger('log', 'Change in title!');
  // Check the URL, ensure that it's a "video"
  const currentHref = window.location.href;

  if (currentHref.indexOf('/watch?v=') >= 0) {
    mutationSet = false, cContent = false, withinComments = ''; // eslint-disable-line
    const video = document.querySelector('#container .html5-video-container > video.video-stream');

    if (!activeTimeUpdate) {
      activeTimeUpdate = true;
      video.addEventListener('timeupdate', timeUpdate);

      function timeUpdate() { // eslint-disable-line
        // Check for current track
        const boolPara = !timeStampObj.links_tracks.length;

        if (Object.keys(timestampHistory.history).length) {
          const historyKeys = Object.keys(timestampHistory.history);

          if (historyKeys.indexOf(window.location.href.substring(0, 43)) >= 0) {
            cContent = document.querySelector('#comments #sections > #contents').children[timestampHistory.history[window.location.href.substring(0, 43)]];
          }
        }

        const timeObj = findCurrentTimestamps(Math.floor(video.currentTime), boolPara, timeStampObj.links_tracks, cContent);

        if (timeObj) {
          timeStampObj = timeObj;
        } else {
          const ytContent = document.querySelector('#yt_timestamp_container');

          if (ytContent !== null) {
            ytContent.parentNode.removeChild(ytContent); // Remove appended content
            if (ytChapterContainer.length) {
              ytChapterContainer[0].textContent = '';
              document.querySelector('div.ytp-chapter-container').style.display = 'none';
            }
          }

          if (withinComments === false) {
            activeTimeUpdate = false;
            video.removeEventListener('timeupdate', timeUpdate);
          }
        }
      };
    }
  }
});

targetObserve.observe(targetEle, {
  childList: true, // # DEV-NOTE Possible replace this with characterData && characterDataOldValue
});

function findCurrentTimestamps(currentVideoTime = 0, searchForTimestamps = true, timestampsCurrent = [], commentContent = false) {
  // Check if any timestamps currently exist
  let descriptionContent = document.querySelector('#description > .content');

  // Check if video has YouTube Chapter support [https://github.com/TylerJDev/YouTube-Timestamps/issues/19]
  const ytChapter = document.querySelector('div.ytp-chapter-container div.ytp-chapter-title');

  if (ytChapter !== null && document.querySelector('#yt_timestamp_container') !== null) {
    const timestampContain = ytChapter.querySelectorAll('span:not(.ytp-chapter-title-prefix):not([aria-hidden])');

    // If valid timestamp title is present
    if (timestampContain !== null && timestampContain.length && timestampContain[0].textContent.length && !timestampContain[0].classList.contains('yt_timestamp_active_text')) {
      consoleLogger('log', timestampContain[0].textContent);
    } else if (timestampContain !== null && timestampContain.length && !timestampContain[0].textContent.length) {
      ytChapterContainer = [timestampContain[0]];

      timestampContain[0].classList.add('yt_timestamp_active_text');

      if (document.querySelector('div.ytp-chapter-container').getAttribute('style').includes('display: none;')) {
        document.querySelector('div.ytp-chapter-container').style.display = 'block';
      }
    }
  }


  if (commentContent !== false) {
    descriptionContent = commentContent;
  }

  if (descriptionContent.textContent !== currentDescr) {
    searchForTimestamps = true;
    currentDescr = descriptionContent.textContent;
    timestampObj = { links_tracks: [] };
    timestampsCurrent = [];
  }

  if (!descriptionContent.querySelectorAll('a').length) {
    return commentContent === false ? checkCommentsForTimestamps() : false; // If no links (timestamps) are found in the description
  }

  const currentLinks = Array.from(descriptionContent.querySelectorAll('a'));
  if (searchForTimestamps === true) {
    // Grab the corresponding timestamp text
    currentLinks.forEach((curr, index) => {
      const val = timestampToSeconds(curr.textContent);
      if (val >= 0 && typeof val === 'number') {
        curr.classList.add('yt_timestamp_link'); // Add class to proper timestamp link

        // Check if link is not unique
        const this_href = curr.attributes.href.value;
        const duplicateLinks = descriptionContent.querySelectorAll(`a[href="${this_href}"]`);

        timestampsCurrent.push([curr, grabTimestampText(curr, descriptionContent, this_href, duplicateLinks)]); // timestampToSeconds returns an integer
      }
    });
  }

  // Grab the current track based on currentVideoTime parameter passed
  let currentTimestamp = timestampsCurrent.filter((currEle) => currEle[0].textContent === determineTimeSlot(currentVideoTime, timestampsCurrent.map((curr) => curr[0].textContent)));

  if (currentTimestamp.length === 2) {
    console.warn('2 non-unique timestamps found!');
    currentTimestamp = [currentTimestamp[1]];
  }

  if (currentTimestamp.length) {
    if (commentContent !== false && timestampsCurrent.length <= 4) { // If searching through comment(s) and links found is less or equal to 4
      return false;
    }

    const timestampObj = {
      links_tracks: timestampsCurrent, current_link: currentTimestamp[0][0], current_track: currentTimestamp[0][1], current_video: window.location.href, current_index: '',
    };

    if (isNaN(timestampToSeconds(timestampObj.links_tracks[0][0].textContent.trim()))) { // If the first element in links_track text !== number
      return false;
    }

    timestampObj.links_tracks.forEach((x, index) => {
      if (x[0] === timestampObj.current_link) {
        timestampObj.current_index = index;
      }
    });

    if (!timestampObj.current_link.classList.contains('selected_yt_timestamp_link')) {
      // Remove aria-describedby
      Array.from(document.querySelectorAll('a.selected_yt_timestamp_link'), (curr) => curr.removeAttribute('aria-describedby'));

      // Remove previous selected classes
      Array.from(document.querySelectorAll('a.selected_yt_timestamp_link'), (curr) => curr.classList.remove('selected_yt_timestamp_link'));
    }

    const hideBg = settings.disable_background === true ? '' : 'yt_timestamps_bg_hide';

    // Add class to current link
    timestampObj.current_link.classList.add('selected_yt_timestamp_link');
    if (hideBg.length) {
      timestampObj.current_link.classList.add(hideBg);
    }

    if (document.querySelector('a.selected_yt_timestamp_link') !== null) {
      document.querySelector('a.selected_yt_timestamp_link').setAttribute('aria-describedby', 'yt_timestamp_title_playing');
    }

    // Create heading if it doesn't exist
    const headingTimestampTitle = document.querySelector('h2.yt_timestamp_nowplaying');

    // Create Next/Previous buttons
    const nextPrevButtons = document.querySelector('button.yt_timestamp_controls');

    function changeByControl(currEle) {
      const currentControlType = this.attributes['data-controltype'].value;

      if (currentControlType === 'next') {
        document.querySelector('#container .html5-video-container > video.video-stream').currentTime = timestampToSeconds(timeStampObj.links_tracks[timeStampObj.current_index + 1][0].textContent);
      } else if (currentControlType === 'previous') {
        const count = timestampToSeconds(timeStampObj.current_link.textContent) === timestampToSeconds(timeStampObj.links_tracks[timeStampObj.current_index - 1][0].textContent) ? 2 : 1;

        document.querySelector('#container .html5-video-container > video.video-stream').currentTime = timestampToSeconds(timeStampObj.links_tracks[timeStampObj.current_index - count][0].textContent);
      }
    }

    const hideBtn = timeStampObj.current_index === 0 ? '.yt_timestamp_prev' : timestampObj.current_index + 1 === timeStampObj.links_tracks.length ? '.yt_timestamp_next' : false;

    if (headingTimestampTitle === null) {
      const hideHeading = settings.toggle_heading === true ? '' : 'yt_timestamps_hide';
      const containerPosition = settings.below_title === true ? 'afterend' : settings.above_title === true ? 'beforebegin' : 'afterend';

      // Create a container
      document.querySelector('#container > h1.title').insertAdjacentHTML(containerPosition, '<div id="yt_timestamp_container"></div>');

      document.querySelector('#yt_timestamp_container').insertAdjacentHTML('afterbegin', `<h2 id="yt_timestamp_title_playing" class="${`${hideHeading} ${hideBg}`} title style-scope ytd-video-primary-info-renderer yt_timestamp_nowplaying" aria-live="polite" role="status">Now Playing: ${timestampObj.current_track}</h2>`);
      if (nextPrevButtons === null) { // Double check
        document.querySelector('h2.yt_timestamp_nowplaying').insertAdjacentHTML('beforebegin', `<button class="${hideBg} yt_timestamp_controls yt_timestamp_prev" aria-label="Previous Track" data-controltype="previous"> &#9664; </button>`);
        document.querySelector('h2.yt_timestamp_nowplaying').insertAdjacentHTML('afterend', `<button class="${hideBg} yt_timestamp_controls yt_timestamp_next" aria-label="Next Track" data-controltype="next"> &#9658; </button>`);
        Array.from(document.querySelectorAll('button.yt_timestamp_controls'), (x) => x.addEventListener('click', changeByControl));

        // Reuse next/prev buttons
        const nextPrev = [document.querySelector('.ytp-left-controls .ytp-next-button svg'), document.querySelector('.ytp-left-controls .ytp-prev-button svg')];
        if (nextPrev.indexOf(null) === -1) {
          const nextPrevElements = [document.querySelector('.yt_timestamp_next'), document.querySelector('.yt_timestamp_prev')];
          nextPrevElements.forEach((curr, index, arr) => {
            curr.textContent = '';
            curr.insertAdjacentHTML('afterbegin', nextPrev[index].outerHTML);
          });
        }

        defaultStyles();
      }
    } else if (headingTimestampTitle.textContent !== timeStampObj.current_track) {
      // Add title to heading
      headingTimestampTitle.textContent = timestampObj.current_track;

      if (ytChapterContainer.length) {
        ytChapterContainer[0].textContent = timestampObj.current_track;
      }
    }

    // If video does not contain timestamps, ensure timeStampObj is empty, only with key 'links_track'
    if (timeStampObj.current_link !== undefined && (timeStampObj.current_link.parentNode === null || timeStampObj.current_link.href.substring(0, 43) !== window.location.href.substring(0, 43))) {
      if (timestampObj.current_track !== timeStampObj.current_track) {
        timeStampObj = timestampObj;
        return timestampObj;
      }

      timeStampObj = { links_tracks: [] };
      return false;
    }

    // Adjust color based on current settings within {settings} global
    if (document.querySelector('#yt_timestamp_container').children.length
       && (document.querySelector('#yt_timestamp_container').children[0].getAttribute('style') !== null
       && document.querySelector('#yt_timestamp_container').children[0].getAttribute('style').indexOf(`${settings.color.trim()} !important`) === -1)
    ) {
      consoleLogger('log', 'Color has been changed', 'to', settings.color.trim());
      Array.from(document.querySelector('#yt_timestamp_container').children, (currElem) => currElem.setAttribute('style', `background-color: ${settings.color.trim()} !important`));
    }

    if (hideBtn !== false
       && (document.querySelector(hideBtn) !== null && document.querySelector(hideBtn).classList.contains('yt_disabled_btn') === false)) {
      consoleLogger('log', 'yt_disabled_btn has been added to:', hideBtn);
      document.querySelector(hideBtn).classList.add('yt_disabled_btn');
    } else if (hideBtn === false && document.querySelector('button.yt_disabled_btn') !== null) {
      consoleLogger('log', 'yt_disabled_btn has been removed');
      document.querySelector('button.yt_disabled_btn') !== null ? document.querySelector('button.yt_disabled_btn').classList.remove('yt_disabled_btn') : '';
    }

    return timestampObj;
  }
  return commentContent === false ? checkCommentsForTimestamps() : false;
}

function grabTimestampText(ele, descr, href, nonUnique) {
  const siblings = [ele.previousSibling, ele.nextSibling]; // [0] = Some random text not related, [1] Proper timestamp text
  const descrText = descr.textContent.split('\n');
  let defTitle = '';

  const timestampText = descrText.filter((curr, idx) => {
    // @ If previous sibling exists, and this current line within the description has the previous sibling's text content as well as timestamp which was passed in this function
    if (siblings[0] !== null && curr.indexOf(siblings[0].textContent.trim()) >= 0 && curr.indexOf(ele.textContent) >= 0 && ele === nonUnique[nonUnique.length - 1]) {
      return curr;
    } if (siblings[1] !== null && curr.indexOf(siblings[1].textContent.trim().split('\n')[0]) >= 0 && curr.indexOf(ele.textContent) >= 0) {
      defTitle = curr;
      return curr;
    } if (siblings[0] !== null
      && (siblings[0].textContent.split('\n').length > 1 && curr.indexOf(siblings[0].textContent.trim().split('\n').pop()) >= 0 && curr.indexOf(ele.textContent) >= 0)) {
      return siblings[0].textContent.split('\n').pop();
    } if (siblings[1] !== null
      && (siblings[1].textContent.split('\n').length > 1 && curr.indexOf(siblings[1].textContent.trim().split('\n').pop()) >= 0)) {
      return siblings[1].textContent.split('\n').pop();
    } if (curr.indexOf(ele.textContent) >= 0 && curr.replace(ele.textContent).trim().length) {
      return curr.replace(ele.textContent).trim();
    }
  });

  if (nonUnique.length > 1 && ele === nonUnique[nonUnique.length - 1] && timestampText.length === 2) {
    timestampText.shift();
  }

  if (timestampText.length <= 5) {
    return parseString(timestampText[0], ele.textContent);
  }
  return parseString(defTitle, ele.textContent);
}

/**
* Parse Timestamp Text
*
* @param {string} str - The string to parse
* @param {string} content - The original timestamp itself ("0:00")
*
* @example
*
*   parseString('1:00 Best Track Ever', '1:00')
*/

const parseString = (str, content) => {
  const hasTimestamp = str.replace(/[\[\]()]+/g, '').split(' ').filter((currStr) => timestampToSeconds(currStr));

  if (hasTimestamp.length) {
    for (const foundStamps in hasTimestamp) {
      str = str.replace(hasTimestamp[foundStamps], '');
    }
  }

  return str.replace(content, '')
    .replace(/[()\[\]-] +/gi, '')
    .replace(':', '').trim();
};

function timestampToSeconds(timestamp) {
  // If valid timestamp, i.e, timestamp.split(':') => ['00', '00', '00'].length = 3
  if (timestamp !== undefined && timestamp.split(':').length <= 3 && timestamp.indexOf(':') >= 0) {
    timestamp = timestamp.replace(/[:]/gi, '');

    // timestamp = 2:00, needs to be 00:02:00
    const currentTcount = (6 - timestamp.length);

    timestamp = currentTcount > 0 ? `${'0'.repeat(currentTcount)}${timestamp}`.match(/.{0,2}/g) : timestamp.match(/.{0,2}/g);
    timestamp.pop();
    timestamp = timestamp.join(':');

    const defTime = new Date(`Jan 1, 70 ${timestamp} GMT+00:00`);
    return defTime.getTime() / 1000;
  }

  return false; // Default, if clause didn't pass
}

function determineTimeSlot(time, tracks) { // 65, ['00:03', '2:38', '4:00'];
  // Current time in seconds
  const currTime = time; // 65

  // Filter through the entire timestamps array, if current timestamp time is less than "this" timestamp time ...
  const trackTypes = tracks.filter((trackTime) => currTime < timestampToSeconds(trackTime));

  if (!trackTypes.length && currTime > timestampToSeconds(tracks[tracks.length - 1])) { // If last track
    return tracks.filter((trackTime) => currTime > timestampToSeconds(trackTime[0])).reverse()[0];
  }

  // Find the exact timestamp that matches the above var "trackTypes"
  const currentSlot = tracks[tracks.indexOf(trackTypes[0]) - 1]; // Find index of trackTypes, get the array element (index) before this

  if (currentSlot === undefined && currTime < timestampToSeconds(trackTypes[0])) { // This is for a specific issue, @ watch?v=_MVcJDzX-OU
    return trackTypes[0]; // If no timestamp is found (currentSlot) and current video time is LESS than first trackTypes item, return first trackTypes item
  }

  return currentSlot;
}

function checkCommentsForTimestamps() {
  const titleObserve = new MutationObserver(((currMutation) => {
    if (document.querySelector('ytd-comments#comments #contents').childElementCount > 0) {
      const setupCommentCheck = new Promise((resolve, reject) => {
        setTimeout(() => {
          resolve(document.querySelector('ytd-comments#comments #contents').children);
        }, 1500);
      });

      setupCommentCheck.then((result) => {
        const toArray = Array.from(result);
        for (const currNode in toArray) {
          if (findCurrentTimestamps(Math.floor(document.querySelector('#container .html5-video-container > video.video-stream').currentTime), true, [], toArray[currNode]) !== false) {
            cContent = toArray[currNode];
            timestampHistory.history[window.location.href.substring(0, 43)] = Array.from(toArray[currNode].parentNode.children).indexOf(toArray[currNode]);

            break;
          }
        }

        if (cContent === false) { // If no timestamps are found within the comments
          withinComments = false;
        }
      });

      titleObserve.disconnect();
    }
  }));

  if (document.querySelector('ytd-comments#comments #contents') !== null && document.querySelector('ytd-comments#comments #contents').childElementCount === 0 && mutationSet === false) {
    titleObserve.observe(document.querySelector('ytd-comments#comments #contents'), {
      childList: true,
    });
    mutationSet = true;
  } else {
    return false;
  }
}

function runSettings(currSettings) {
  const changedSetting = Object.keys(currSettings.plugin_settings.oldValue).filter((curr, idx) => {
    if (currSettings.plugin_settings.oldValue[curr] !== currSettings.plugin_settings.newValue[curr]) {
      return curr;
    }
  });

  consoleLogger('log', currSettings);
  consoleLogger('log', changedSetting);

  if (changedSetting.indexOf('toggle_heading') > -1) {
    settings.toggle_heading === true
      ? document.querySelector('#yt_timestamp_title_playing').classList.remove('yt_timestamps_hide')
      : document.querySelector('#yt_timestamp_title_playing').classList.add('yt_timestamps_hide');
  }

  if (changedSetting.indexOf('disable_background') > -1) {
    settings.disable_background === true
      ? Array.from(document.querySelectorAll('h2.yt_timestamp_nowplaying, button.yt_timestamp_controls, a.selected_yt_timestamp_link'), (x) => x.classList.remove('yt_timestamps_bg_hide'))
      : Array.from(document.querySelectorAll('h2.yt_timestamp_nowplaying, button.yt_timestamp_controls, a.selected_yt_timestamp_link'), (x) => x.classList.add('yt_timestamps_bg_hide'));
  }

  if (changedSetting.indexOf('above_title') > -1 || changedSetting.indexOf('below_title') > -1) {
    if (settings.below_title) {
      document.querySelector('#info #container').insertBefore(document.querySelector('#yt_timestamp_container'), document.querySelector('#container > h1.title').nextSibling);
    } else if (settings.above_title) {
      document.querySelector('#info #container').insertBefore(document.querySelector('#yt_timestamp_container'), document.querySelector('#container > h1.title'));
    }
  }

  if (changedSetting.indexOf('color') > -1) {
    // Check if current color !== settings.color
    Array.from(document.querySelector('#yt_timestamp_container').children, (currElem) => currElem.setAttribute('style', `background-color: ${settings.color} !important`));
  }
}

chrome.storage.sync.get(['plugin_settings'], (result) => {
  if (Object.keys(result).length) {
    settings = result.plugin_settings;
  }
});

chrome.storage.onChanged.addListener((changes) => {
  settings = changes.plugin_settings.newValue;
  runSettings(changes);
});

try {
  module.exports = {
    timestampToSeconds, determineTimeSlot, findCurrentTimestamps,
  };
} catch (ReferenceError) {
  consoleLogger('log', 'temp');
}

// eslint-disable-next-line no-unused-vars
function defaultStyles(type) {
  let newType = type;

  // Grab from localstorage
  const storedValue = localStorage.getItem('yt_timestamp_style');

  if (storedValue !== null && typeof newType !== 'boolean') {
    newType = storedValue === 'true' ? true : storedValue === 'false' ? false : '';
  }

  if (newType === true) {
    document.querySelectorAll('.yt_timestamp_nowplaying, .yt_timestamp_controls').forEach((curr) => {
      curr.classList.add('yt_timestamp_default_style');
    });
  } else if (newType === false) {
    document.querySelectorAll('.yt_timestamp_nowplaying, .yt_timestamp_controls').forEach((curr) => {
      curr.classList.remove('yt_timestamp_default_style');
    });
  }

  localStorage.setItem('yt_timestamp_style', typeof newType !== 'boolean' ? false : newType);
}
